### PR TITLE
Feat: Autopeering without public IP

### PR DESCRIPTION
**Pull request by Wollac**
- The peering end point of a peer is the UDP envelope address + provided peering port
- Services only describes port on the peer IP
- Check that the port of the peering service matches the port the ping was sent to

### CODE REVIEW

File: autopeering/discover/protocol.go [hunk 1]
```diff
@@ -364,50 +400,46 @@ func (p *Protocol) validatePing(fromAddr string, m *pb.Ping) bool {
...
 }
 
-func (p *Protocol) handlePing(s *server.Server, fromAddr string, fromID peer.ID, fromKey peer.PublicKey, rawData []byte) {
+func (p *Protocol) handlePing(s *server.Server, fromAddr *net.UDPAddr, fromID peer.ID, fromKey peer.PublicKey, m *pb.Ping, rawData []byte) {
 	// create and send the pong response
 	pong := newPong(fromAddr, rawData, p.loc.Services().CreateRecord())
 
-	p.logSend(fromAddr, pong)
-	s.Send(fromAddr, marshal(pong))
+	// the destination address takes uses the source IP address of the packet plus the src_port in the message
```
Comment (muXxer) [inline autopeering/discover/protocol.go]: there is a word too much in this sentence.... "takes uses"

File: autopeering/peer/service/record.go [hunk 2]
```diff
@@ -11,34 +10,39 @@ import (
...
+	port    int
 }
 
 // Network returns the service's network name.
-func (a *networkAddress) Network() string {
+func (a endpoint) Network() string {
 	return a.network
 }
 
 // String returns the service's address in string form.
```
Comment (muXxer) [inline autopeering/peer/service/record.go]: wrong comment

File: autopeering/discover/protocol.go [hunk 3]
```diff
@@ -32,36 +34,33 @@ var retryPolicy = backoff.ExponentialBackOff(500*time.Millisecond, 1.5).With(
 type Protocol struct {
 	server.Protocol
 
-	loc *peer.Local    // local peer that runs the protocol
-	log *logger.Logger // logging
+	version uint32         // specifies the expected version number for this Protocol
```
Comment (capossele) [inline autopeering/discover/protocol.go]: Should we explicitly split the `version` field into networkID and version? WDYT?

Comment (muXxer) [reply autopeering/discover/protocol.go]: Yes :)

Comment (Wollac) [reply autopeering/discover/protocol.go]: So have a `uint32 version` and a `string networkID` or what are you proposing?

Comment (capossele) [reply autopeering/discover/protocol.go]: the version field could still be there just for defining the protocol version. Maybe 1 byte is enough for that? 256 possible different versions? For the networkID I like the idea of using string but it should be short and of fixed size

**Pull request merged**